File: 34-find-first-and-last-position-of-element-in-sorted-array/find-first-and-last-position-of-element-in-sorted-array.py
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        left=self.binSearch(nums,target,True)
        right=self.binSearch(nums,target,False)
        return [left,right]
    def binSearch(self,nums,target,status):
       
        left = 0
        right = len(nums) - 1
        idx=-1

        while left <= right:
            mid = (left + right) // 2

            if nums[mid] == target:
                idx=mid
                if status:
                    right=mid-1
                else:
                    left=mid+1

            elif nums[mid] < target:
                left = mid + 1
            else:
                right = mid - 1

        return idx



#print(Solution.searchRange(Solution, [5, 7, 7, 8, 8, 10], 7))

File: 34-find-first-and-last-position-of-element-in-sorted-array/find-first-and-last-position-of-element-in-sorted-array.cpp
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        int left = binSearch(nums, target, true);
        int right = binSearch(nums, target, false);
        return {left, right};
    }

private:
    int binSearch(vector<int>& nums, int target, bool findFirst) {
        int left = 0;
        int right = nums.size() - 1;
        int idx = -1;

        while (left <= right) {
            int mid = left + (right - left) / 2;
            
            if (nums[mid] == target) {
                idx = mid;
                if (findFirst) {
                    right = mid - 1;
                } else {
                    left = mid + 1;
                }
            } else if (nums[mid] < target) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        return idx;
    }
};
